### PR TITLE
Implement Regression Test and Capture Baseline

### LINKED ISSUE #6: Create a regression test baseline
# Create Regression Test Baseline

## Overview

Create a comprehensive regression test suite with golden master baselines to serve as a safety net before refactoring the petal control codebase. This establishes objective, automated verification that refactored code produces identical results to the current implementation.

## Motivation

- **No existing tests**: Current codebase has ~0% test coverage
- **High-risk refactoring**: Planned restructuring of 12,000+ lines across 18 modules
- **Critical system**: Controls physical hardware; bugs have real-world consequences
- **Complex behavior**: Collision detection, scheduling, and transforms have subtle interdependencies

Having regression tests provides:
- Confidence to make aggressive changes
- Immediate detection of unintended behavior changes
- Living documentation of system behavior
- Objective measure of refactoring completion

### 1. Regression Test Framework (`regression_test.py`)

Core infrastructure for capturing and comparing system behavior:

- [ ] Golden master baseline serialization (JSON + SHA256 signatures)
- [ ] Floating-point tolerant comparison (6 decimal places)
- [ ] Detailed diff reporting when tests fail
- [ ] Baseline save/load/update workflow
- [ ] Test runner with summary reporting

### 2. Test Coverage (Minimum 8 Test Suites)

- [ ] **Basic Moves**: All coordinate systems (posintTP, poslocTP, poslocXY, ptlXY, QS, obsXY)
- [ ] **Collision Scenarios**: Known collision cases using existing harness sequences
- [ ] **Edge Cases**: Theta wrap-around, phi limits, retracted positioners, locked axes
- [ ] **State Management**: Save/load persistence, state updates, validation
- [ ] **Transform Chain**: Forward/reverse/roundtrip for all coordinate systems
- [ ] **Scheduling Algorithms**: All anticollision modes × anneal modes
- [ ] **Move Table Formats**: All 7 output formats (schedule, collider, hardware, etc.)
- [ ] **Complex Sequences**: Multi-move with corrections, mixed coordinates, partial/full petal

### 3. Integration with Existing Harness

Minimal modifications to `anticollision_test_harness/harness.py`:

- [ ] Add result capture class (`HarnessResult`)
- [ ] Record move execution details (requests, tables, frozen positioners)
- [ ] Export results in regression-test compatible format
- [ ] Maintain backward compatibility with existing usage

### 4. Baseline Data

- [ ] Generate and commit initial baselines (`regression_baselines/*.json`)
- [ ] Document baseline creation date and code version
- [ ] Verify baselines are reproducible (run twice, same hash)

### 5. Documentation

- [ ] README explaining how to run tests (`--mode baseline` vs `--mode compare`)
- [ ] CI/CD integration instructions
- [ ] Guidelines for updating baselines when behavior intentionally changes
- [ ] Known issues tracking (bugs captured as "correct" baseline)

### Golden Master Strategy

1. **Capture**: Run system with representative inputs, serialize all outputs
2. **Hash**: Compute SHA256 signature of serialized data
3. **Compare**: On subsequent runs, compare signatures (fast) or data (detailed)
4. **Report**: Show exactly what changed and where

### State Capture

For each test scenario, record:
- Final positioner positions (all coordinate systems)
- Move table contents (timing, distances, pauses)
- Collision detection results (frozen positioners, collision cases)
- Scheduling statistics (methods used, success rates)
- Transform roundtrip accuracy

### Tolerance Handling

- Round floats to 6 decimal places before comparison
- Avoid timing-dependent behavior (use fixed seeds for random)
- Normalize dict/set ordering in serialization

## Acceptance Criteria

- [x] All 8+ test suites implemented and passing
- [x] Baselines committed to git (`regression_baselines/`)
- [x] Tests run in < 5 minutes on standard hardware
- [ ] `make test` or `python regression_test.py` runs all tests
- [x] Zero failures when run against committed baselines
- [x] Documentation complete and reviewed

## Effort Estimate

**2 weeks** (1 developer):
- Week 1: Framework + first 4 test suites + harness integration
- Week 2: Remaining 4 test suites + documentation + baseline verification

## Dependencies

- Existing `anticollision_test_harness/` infrastructure
- Representative test sequences (already exist in `sequences.py`)
- Python 3.7+ with json, hashlib (standard library)

## Follow-up Work

After completing this work (Phase 0):
- **Phase 1**: Refactor PosState (with regression tests validating)
- **Phase 2**: Break up petal.py (with regression tests validating)
- **Phase 3+**: Continue migration strategy with safety net in place

## Notes

- Focus on **behavioral correctness**, not implementation details
- Tests should be **black-box** where possible (inputs → outputs)
- **Coarse-grained** initially; add fine-grained tests as needed
- Mark known bugs/issues in baseline metadata for future fix validation
- Keep test data in git (helps with debugging when tests fail)

## Related Issues

- #TBD: Refactor PosState (blocks: this issue)
- #TBD: Break up petal.py God object (blocks: this issue)
- #TBD: Extract coordinate transform hierarchy (blocks: this issue)

### LINKED ISSUE #9: Create Minimal fp_settings for Testing
Based on my analysis of the regression test code and its dependencies, here are the **required top-level subdirectories** from fp_settings:

## Strictly Required

**1. `pos_settings/`**
- **Purpose**: Stores positioner configuration files
- **Required file**: `_unit_settings_DEFAULT.conf` (7.8K)
- **Why**: Used by `PosState.load_from_cfg()` as a template when creating new positioner configs
- **Runtime behavior**: Tests create individual `unit_*.conf` files for test positioners

**2. `collision_settings/`**
- **Purpose**: Anti-collision algorithm parameters
- **Required file**: `_collision_settings_DEFAULT.conf` (4.0K)  
- **Why**: Used by `PosCollider.__init__()` when `collider_file=None` (which is the regression test default)
- **Code reference**: `poscollider.pyx:29-32`

## Optional (referenced in code but not used by regression tests)

**3. `fid_settings/`**
- Contains: `_unit_settings_DEFAULT.conf` (2.3K)
- Not required because regression tests use `fidids={}` (empty)

**4. `ptl_settings/`**
- Contains: `_unit_settings_DEFAULT.conf` (1.1K)
- Not required because regression tests explicitly provide `posids` parameter

**5. `test_settings/`**
- Contains: Various xytest data files
- Not accessed by regression tests

**6. `other_settings/`**
- Contains: Power supply configuration files
- Not accessed by regression tests

**7. `hwsetups/`**
- Contains: Hardware setup configs, XY location files
- Defined in `posconstants.py` but not actually used by core code

## Summary

**Minimal fp_settings directory structure for regression tests:**
```
fp_settings/
├── pos_settings/
│   └── _unit_settings_DEFAULT.conf
└── collision_settings/
    └── _collision_settings_DEFAULT.conf
```

Total size: ~12KB for the two required files.

### PR DESCRIPTION
**Pull request by dkirkby**
Fixes #6 Fixes #9

### CODE REVIEW

**Comment by dkirkby** [comment]
These are the diffs in the production code required to get the regression tests working, generated with:
```bash
git diff main..regtest -- 'petal/' ':(exclude)petal/regression/' ':(exclude)petal/.coveragerc'
```

### petal.py

Improved error reporting when `import poscollider` fails (could be due to a cython error or a problem with something imported by poscollider.pyx).

```diff
diff --git a/petal/petal.py b/petal/petal.py
index 68edf3af..03504c3e 100644
--- a/petal/petal.py
+++ b/petal/petal.py
@@ -4,10 +4,14 @@ import posmovetable
 import posstate
 try:
     import poscollider
-except:
+except Exception as e:
     import sys
+    import traceback
     print('Error while importing poscollider. Make sure you have compiled the cython file ' +
           'poscollider.pyx. Use setup.py in the petal directory, see usage comments in there.')
+    print(f'\nActual error: {type(e).__name__}: {e}')
+    print('\nFull traceback:')
+    traceback.print_exc()
     sys.exit(1)
 import posconstants as pc
 import posschedstats
```

### posmodel.py

Assume motor is not linphi if ZENO_MOTOR_P is not defined.

```diff
diff --git a/petal/posmodel.py b/petal/posmodel.py
index 8487dd89..5740426a 100644
--- a/petal/posmodel.py
+++ b/petal/posmodel.py
@@ -27,7 +27,7 @@ class PosModel(object):
         self._stepsize_creep             = 0.1    # deg
         self._motor_speed_cruise         = {pc.T: 9900.0 * 360.0 / 60.0, pc.P: 9900.0 * 360.0 / 60.0} # deg/sec (= RPM *360/60)
         self._stepsize_cruise            = {pc.T: 3.3, pc.P: 3.3} # deg/step
-        if self.state._val['ZENO_MOTOR_P'] is True:
+        if self.state._val.get('ZENO_MOTOR_P', False) is True:
             if self.DEBUG:
                 self.printfunc(f'PosModel: new linphi posid = {posid}')  # DEBUG
 #           self.linphi_params['LAST_P_DIR'] = 1    # 1 is CCW, -1 is CW
@@ -39,10 +39,10 @@ class PosModel(object):
 
     @property
     def is_linphi(self):
-        return self.state._val['ZENO_MOTOR_P'] is True
+        return self.state._val.get('ZENO_MOTOR_P', False) is True
 
     def get_zeno_scale(self, which):    # specify 'SZ_CW_P', 'SZ_CCW_P', or the _T varieties
-        scale = self.state._val[which]
+        scale = self.state._val.get(which)
         if scale is None:
             scale = 1.0
         return scale
```

### posschedule.py

Assume motor is not linphi if ZENO_MOTOR_P is not defined. Ensure that logging works when stats are not enabled.

```diff
diff --git a/petal/posschedule.py b/petal/posschedule.py
index 5d79ce94..e575ccda 100644
--- a/petal/posschedule.py
+++ b/petal/posschedule.py
@@ -295,7 +295,7 @@ class PosSchedule(object):
         colliding_posids = [posid for posid in colliding_sweeps]
         for posid in colliding_posids:
             p_state = self.petal.posmodels[posid].state
-            if p_state._val['ZENO_MOTOR_P'] is True:
+            if p_state._val.get('ZENO_MOTOR_P', False) is True:
                 zeno_posids.add(posid)
         if zeno_posids:
             colliding = set(colliding_sweeps)
@@ -997,8 +997,8 @@ class PosSchedule(object):
 
     def _schedule_moves_initialize_logging(self, anticollision):
         """Initial logging tasks for the schedule_moves() function."""
+        self.__timer_start = time.perf_counter()
         if self.stats.is_enabled():
-            self.__timer_start = time.perf_counter()
             self.stats.set_scheduling_method(str(anticollision))
             self.__original_request_posids = set(self._requests.keys())
             self.__max_net_time = 0
```

### setup.py

Update cython build script and instructions to be compatible with python 3.13 where distutils is deprecated. Prefer setuptools instead.

```diff
diff --git a/petal/setup.py b/petal/setup.py
index cfe81256..e723b71f 100644
--- a/petal/setup.py
+++ b/petal/setup.py
@@ -1,4 +1,7 @@
-from distutils.core import setup
+try:
+    from setuptools import setup
+except ImportError:
+    from distutils.core import setup
 from Cython.Build import cythonize
 #import os
 
@@ -12,10 +15,14 @@ for module in ['poscollider']:
 # General tips for the Cython-uninitiated...
 # ------------------------------------------
 # sample call to compile at general console:
-# python setup.py build_ext --inplace clean --all
+# python setup.py build_ext --inplace
+#
+# If you have issues (e.g., after switching Python versions), clean first:
+# python setup.py clean --all
+# python setup.py build_ext --inplace
 #
 # sample call to compile in Spyder console (generally, do so within a fresh console):
-# runfile('setup.py', args='build_ext --inplace clean --all')
+# runfile('setup.py', args='build_ext --inplace')
 #
 # for Windows compiling, need build tools for visual studio:
 # https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2017
```
